wgpu-core/src/indirect_draw_validation.rs

```diff
@@ -0,0 +1,1080 @@
+                    1,
+                    Some(src_bind_group),
+                    &[batch.src_dynamic_offset as u32],
+                );
+            }
+
+            let dst_bind_group = resources.get_dst_bind_group(batch.dst_resource_index);
+            unsafe {
+                encoder.set_bind_group(pipeline_layout, 2, Some(dst_bind_group), &[]);
+            }
+
+            unsafe {
+                encoder.dispatch([(batch.entries.len() as u32).div_ceil(64), 1, 1]);
+            }
+        }
+
+        unsafe {
+            encoder.end_compute_pass();
+        }
+
+        for index in batches
+            .values()
+            .map(|batch| batch.dst_resource_index)
+            .filter(|index| buffer_index_set.insert(*index))
+        {
+            let dst_buffer = resources.get_dst_buffer(index);
+
+            buffer_barriers.push(hal::BufferBarrier {
+                buffer: dst_buffer,
+                usage: hal::StateTransition {
+                    from: wgt::BufferUses::STORAGE_READ_WRITE,
+                    to: wgt::BufferUses::INDIRECT,
+                },
+            });
+        }
+        buffer_index_set.clear();
+
+        unsafe {
+            encoder.transition_buffers(&buffer_barriers);
+        }
+        buffer_barriers.clear();
+
+        for staging_buffer in staging_buffers {
+            temp_resources.push(TempResource::StagingBuffer(staging_buffer));
+        }
+
+        Ok(())
+    }
+
+    pub fn dispose(self, device: &dyn hal::DynDevice) {
+        let IndirectDrawValidation {
+            module,
+            metadata_bind_group_layout,
+            src_bind_group_layout,
+            dst_bind_group_layout,
+            pipeline_layout,
+            pipeline,
+
+            free_indirect_entries,
+            free_metadata_entries,
+        } = self;
+
+        for entry in free_indirect_entries.into_inner().drain(..) {
+            unsafe {
+                device.destroy_bind_group(entry.bind_group);
+                device.destroy_buffer(entry.buffer);
+            }
+        }
+
+        for entry in free_metadata_entries.into_inner().drain(..) {
+            unsafe {
+                device.destroy_bind_group(entry.bind_group);
+                device.destroy_buffer(entry.buffer);
+            }
+        }
+
+        unsafe {
+            device.destroy_compute_pipeline(pipeline);
+            device.destroy_pipeline_layout(pipeline_layout);
+            device.destroy_bind_group_layout(metadata_bind_group_layout);
+            device.destroy_bind_group_layout(src_bind_group_layout);
+            device.destroy_bind_group_layout(dst_bind_group_layout);
+            device.destroy_shader_module(module);
+        }
+    }
+}
+
+fn get_shader_source(supports_indirect_first_instance: bool) -> String {
+    let indirect_first_instance_check = if supports_indirect_first_instance {
+        ""
+    } else {
+        "failed |= first_instance != 0u;"
+    };
+
+    format!(
```

Comment:
Could we move this to a new file - what I've done in the tests is use `str::replace` to replace `{{indirect_first_instance_check}}` in the file source with the check we want. Alternatively, couldn't a pipeline override used with:

```
if disable_indirect_first_instance {
    failed |= first_instance != 0u;
}
```

then DCE should do its thing in the backend.